### PR TITLE
Publish to npm from GitHub actions

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

Closes https://github.com/babel/babel/pull/10663

During GitHub Universe 2020, they presented [environments](https://docs.github.com/en/free-pro-team@latest/actions/reference/environments#referencing-an-environment) with manual approval for GitHub actions: regardless of who has write access to this repository, we can now granularly configure who has access to a given secret.

I gave access to @JLHwung, @existentialism, @hzoo and me.
Currently the stored `NPM_TOKEN` is mine, but it it works we should use @babel-bot's.

Now the publishing process is just this :tada:
```
# Create a new release commit
make new-version

# Push to GitHub
git push --follow-tags
``` 

I'm testing this action in the second commit (I'll revert it before merging), publishing an empty package.
This is what the UI looks like: https://drive.google.com/file/d/1q-kCP7rp-7UKZZfChrI3v-s2iVHi0KDE/view

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12526"></a>

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
This works :tada: :tada: :tada:

Workflow run: https://github.com/babel/babel/actions/runs/430661855
Published package: https://www.npmjs.com/package/@nicolo-ribaudo/babel-publish-test (I released 0.0.1 locally, and 1.0.0 from this action)
GitHub release (if you have write access to the repo): https://github.com/babel/babel/releases/tag/untagged-6839561d8e388548b3b7 - This is completely wrong, but it's just because I didn't push a tag to this PR so the release tool generated the wrong changelog.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -3,11 +3,48 @@ name: Release
...
+  pull_request:
 
 jobs:
-  github_release:
+  log-updates:
+    name: Log packages to publish
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout the new tag
+        uses: actions/checkout@v1.0.0
```
Comment (JLHwung) [inline .github/workflows/release.yml:14]:
```suggestion
        uses: actions/checkout@v2
```
The checkout actions v2 by default do a shallow checkout.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -5,12 +5,43 @@ on:
...
+    environment: npm
+    steps:
+      - name: Checkout the new tag
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Build and Test
+        run: make prepublish
+      - name: Publish to npm
+        run: yarn release-tool publish --yes
```
Comment (nicolo-ribaudo) [inline .github/workflows/release.yml:32]: Note: This line is the only untested step in this PR (I commented it out, run the action, and everything worked). However, we test in in the e2e tests on CircleCI where we publish to verdaccio.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -5,12 +5,43 @@ on:
...
+  npm-release:
+    name: Publish release on npm
+    runs-on: ubuntu-latest
+    needs: log-updates
+    environment: npm
+    steps:
+      - name: Checkout the new tag
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (JLHwung) [inline .github/workflows/release.yml:28]: Q: Is fetch depth 0-based or 1-based?

**Review by JLHwung — APPROVED** [review]

File: .github/workflows/release.yml (hunk 3 above)
Comment (nicolo-ribaudo) [reply .github/workflows/release.yml:28]: It's `1`-based, `0` means "fetch all". We need it in order to find the previous tag.

**Review by hzoo — APPROVED** [review]

**Pull request merged**